src/disk_cache.rs

```diff
@@ -0,0 +1,952 @@
+use bytes::Bytes;
+use bytes::BytesMut;
+use futures::future::BoxFuture;
+use futures::stream;
+use futures::stream::BoxStream;
+use futures::StreamExt;
+use object_store::GetRange;
+use object_store::GetResultPayload;
+use rand::distributions::Alphanumeric;
+use rand::Rng;
+use serde::Deserialize;
+use serde::Serialize;
+use std::ops::Range;
+use std::sync::Arc;
+use tokio::fs;
+use tokio::fs::File;
+use tokio::fs::OpenOptions;
+use tokio::io::AsyncReadExt;
+use tokio::io::AsyncWriteExt;
+
+use object_store::{
+    path::Path, GetOptions, GetResult, ListResult, MultipartUpload, ObjectMeta, ObjectStore,
+    PutMultipartOpts, PutOptions, PutPayload, PutResult,
+};
+
+#[allow(unused)]
+#[derive(Debug, Clone)]
+pub(crate) struct DiskCachedObjectStore {
+    root_folder: std::path::PathBuf,
+    object_store: Arc<dyn ObjectStore>,
+    part_size: usize, // expected to be aligned with mb or kb
+}
+
+impl DiskCachedObjectStore {
+    pub fn new(
+        object_store: Arc<dyn ObjectStore>,
+        root_folder: std::path::PathBuf,
+        part_size: usize,
+    ) -> Self {
+        Self {
+            object_store,
+            root_folder,
+            part_size,
+        }
+    }
+
+    async fn cached_head(&self, location: &Path) -> object_store::Result<ObjectMeta> {
+        let entry = DiskCacheEntry {
+            root_folder: self.root_folder.clone(),
+            location: location.clone(),
+            part_size: self.part_size,
+        };
+        match entry.read_meta().await {
+            Ok(Some(meta)) => Ok(meta),
+            _ => self.object_store.head(location).await,
+        }
+    }
+
+    async fn cached_get_opts(
+        &self,
+        location: &Path,
+        opts: GetOptions,
+    ) -> object_store::Result<GetResult> {
+        let entry = DiskCacheEntry {
+            root_folder: self.root_folder.clone(),
+            location: location.clone(),
+            part_size: self.part_size,
+        };
+
+        let meta = self.maybe_prefetch_range(&entry, &opts.range).await?;
+        let range = self.canonicalize_range(opts.range, meta.size)?;
+        let parts = self.split_range_into_parts(range.clone());
+
+        // read parts, and concatenate them into a single stream. please note that some of these part may not be cached,
+        // we'll still fallback to the object store to get the missing parts.
+        let futures = parts
+            .into_iter()
+            .map(|(part_id, range_in_part)| self.read_part(location, part_id, range_in_part))
+            .collect::<Vec<_>>();
+        let result_stream = stream::iter(futures).then(|fut| fut).boxed();
+
+        Ok(GetResult {
+            meta,
+            range,
+            attributes: Default::default(),
```

Comment:
I don't think we can use the default attributes here. We care about the user-defined attribute `writer_epoch` for fencing the manifest and WAL ssts. And we care about the `compactor_epoch` attribute for fencing compaction SSTs.

Reply:
i've added the cache for attributes, `(meta, attributes)` together is named as `head` in the local cache. (still need more test to cover it though)